### PR TITLE
Added support for timing out on the lambda calls

### PR DESCRIPTION
**Pull request by kf6kjg**
Looked like the existing timeout parameter wasn't being used so I co-opted it.  This prevents each call from taking longer than the given timeout and thus should solve, by masking the underlying problem, most all cases of the warmer never returning and thus incurring huge time penalties.

I added additional output to the final log message so that the timeouts can be observed.

I've tested against an API I've been setting up that uses VPC for the lambdas that had been having issues and warmup was sucking up large blocks of time, 10+ seconds of billable time every 5 minutes, due to a bad config causing it to not work.  Once I got it working, even without this patch, it only takes a couple of seconds to do its job.  With this patch it's possible to set the timeout to 5 seconds and not have to worry about hitting AWS's timeout and thus missing out on log info: you'll only possibly miss success/fail on endpoints that take too long.

An addition to this that I can see would be to make the timeout able to be a per-endpoint setting.

Side note: I did not add protection against the user typing silly things like non-numeric values for the timeout config parameter, as I simply left the inject of that parameter alone.

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -316,10 +317,22 @@ aws.config.region = "${region}";
 const lambda = new aws.Lambda();
 const functions = ${JSON.stringify(functions)};
 
+Promise.delay = function(t, val) {
```
Comment (juanjoDiaz) [inline src/index.js]: Extending built-in types and classes is never a good idea.

File: src/index.js [hunk 2]
```diff
@@ -316,10 +317,22 @@ aws.config.region = "${region}";
...
+  return new Promise(resolve => {
+    setTimeout(resolve.bind(null, val), t);
+  });
+};
+
+Promise.raceAll = function(timeoutTime, timeoutVal, promises) {
+  return Promise.all(promises.map(p => {
```
Comment (juanjoDiaz) [inline src/index.js]: This is the exact same thing as

```js
  return Promise.race([ Promise.delay(timeoutTime, timeoutVal), ...promises ]);
```

but creating a lot more intermediate promises.

File: src/index.js (hunk 1 above)
Comment (kf6kjg) [reply src/index.js]: The way it was done, without any pre-existence check, yes. I blindly C&P'd from SO and that's usually a bad idea.  However I'm game with just making these functions in the current scope or some other better plan.

File: src/index.js (hunk 2 above)
Comment (kf6kjg) [reply src/index.js]: Good catch. Again, my blind C&P from SO was a bad move.  I'll get that cleaned up when I revise this PR.

**Comment by kf6kjg** [comment]
You are correct. And I didn't see that usage, mistaking it for part of the config.  My bad.

As a user I like being informed about which endpoints timed out, and I liked being able to see that the warmer lambda finished correctly instead of just dying. When the function just dies it's hard to tell if it was because of an error in the warmer lambda or that one or more of my lambdas was taking too long to return.

I can definitely look at the `httpOptions` route over the promise race.

So I see a few ways forward, maybe you see some more:
1. Drop this PR, let timeout be only at the warmer level as it has been.  I'm OK with that, won't hurt my feelings: it's your project! :D
1. I can change the meaning of `timeout` to be for the called lambdas and make the warmer lambda time out o some multiplier of that value, e.g. 1.5 times - a simple best-effort attempt to not have the warmer return before the children are all done timing out but still provide a limit to the warmer. I don't really like changing the meaning in this manner as it'll affect people's existing installations.
1. I can modify the PR to add a new per-lambda config parameter that allows the user to figure out what they want on a global or per-lambda level. The default could be set to 10 seconds, and the `timeout` option could have its default moved up to 15 seconds. I choose 10 seconds because it can take over 5 seconds just to get a lambda warmed up enough to return success or error.

What do you think?

**Comment by juanjoDiaz** [comment]
Tackled by #232

**Pull request closed without merging**